lib/service/service.go

```diff
@@ -4219,14 +4229,20 @@ func (process *TeleportProcess) initApps() {
 			return trace.Wrap(err)
 		}
 
+		defer func() {
+			if !ok {
+				warnOnErr(appServer.Close(), log)
```

Comment:
Considering that the whole point of `warnOnErr` is to avoid warning on double close, perhaps we can just `defer func() { warnOnErr(conn.Close()) }` and `defer func() { warnOnErr(appServer.Close()) }` blindly, and not even bother with the `conn.Close` in the `app.shutdown` func?

Reply:
Just to be clear, are you suggesting the same thing as your other comment? (removing the `OnExit` and just deferring the cleanup?)

If not, I'm not sure I understand this suggestion. The point of the deferred close here is only to clean up in the error case (when we don't register the onexit handler).

Reply:
I'm saying that we can just defer the close regardless of whether or not we actually succeed in initializing the app service (and thus close during the shutdown); if you prefer to keep the `ok` flag then that's fine too.

Reply:
But the app service's close is on a separate goroutine, so we'd have to fix that first, otherwise we will prematurely close things, right?

Reply:
Well, `apps.start` will wait until server and agent pool are both closed anyway - it's all very coincidental tho, so I agree that it's better to be explicit with the flag (or with a wait for `TeleportExitEvent`).